src/ert/protobuf/_state_machine.py

```diff
@@ -0,0 +1,239 @@
+        old_real = self._state_handle.get_realization(real.id)
+        # TODO: check the status change is valid (old_real.status, new_real.status)
+        old_real.MergeFrom(real)
+        # TODO: propagate down to jobs if cancel
+
+    @_update.register
+    def _(self, step: pb2.Step) -> None:
+        logger.debug(
+            f"Updating step id {step.id.step} status "
+            f"{pb2.StepStatus.Name(step.status)} from pbuf!",
+        )
+        old_real = self._state_handle.get_step(step.id)
+        # TODO: check the status change is valid (old_real.status, new_real.status)
+        old_real.MergeFrom(step)
+        # if step==success then set realization=success
+        if step.status == pb2.STEP_SUCCESS:
+            self._update(
+                pb2.Realization(id=step.id.realization, status=pb2.STEP_SUCCESS),
+            )
+
+    @_update.register
+    def _(self, ens: pb2.Ensemble) -> None:
+        logger.debug(
+            f"Updating step id {ens.id.id} status"
+            f"{pb2.EnsembleStatus.Name(ens.status)} from pbuf!",
+        )
+        old_ens = self._state_handle.get_ensemble(ens.id)
+        # TODO: check the status change is valid (old_exp.status, new_exp.status)
+        old_ens.MergeFrom(ens)
+
+    @_update.register
+    def _(self, exp: pb2.Experiment) -> None:
+        logger.debug(f"Updating experiment id {exp.id.id} from pbuf!")
+        old_exp = self._state_handle.get_experiment(exp.id)
+        # TODO: check the status change is valid (old_exp.status, new_exp.status)
+        old_exp.MergeFrom(exp)
+
+    async def update(
+        self,
+        msg: Union[
+            pb2.Job,
+            pb2.Step,
+            pb2.Realization,
+            pb2.Ensemble,
+            pb2.Experiment,
+            pb2.DispatcherMessage,
+        ],
+    ) -> None:
+        try:
+            if isinstance(msg, pb2.DispatcherMessage):
+                self._update(getattr(msg, str(msg.WhichOneof("object"))))
+            else:
+                self._update(msg)
+        except Exception:
+            logger.error(f"Failed state machine update! Current state: {self.state}")
+            raise
+
+    @property
+    def state(self) -> pb2.Experiment:
+        assert self._state_handle._experiment is not None
+        return self._state_handle._experiment
+
+
+def node_status_builder(  # pylint: disable=too-many-arguments
+    status: str,
+    experiment_id: str,
+    ensemble_id: Optional[str] = None,
+    realization_id: Optional[int] = None,
+    step_id: Optional[int] = None,
+    job_id: Optional[int] = None,
+) -> pb2.DispatcherMessage:
+    """Builds a DispatcherMessage based on the give argument list.
+    It decides which Protobuf object to create based on the set of ids.
+    It doesn't check whether the status is a valid one, it just raises ValueError
+    in case the status was wrong.
+
+    Args:
+        status (str): get status and converts it protobuf status, validity not checked.
```

Comment:
You can remove the type from these arguments, it is provided by sphinx and typing.

Reply:
(sphinx also has access to the default, so it does not have to be spelled out)